resources/codegen/python/openassetio_codegen/generators/python.py

```diff
@@ -0,0 +1,253 @@
+            subpackage_init_imports = []
+
+            # Generate a single-file module for each namespace
+            for namespace in namespaces:
+                safe_namespace = env.filters["to_py_module_name"](namespace.id)
+                subpackage_init_imports.append(safe_namespace)
+                render_template(
+                    kind,
+                    os.path.join(subpackage_dir_path, f"{safe_namespace}.py"),
+                    {
+                        "package": package_definition,
+                        "namespace": namespace,
+                        "imports": helpers.package_dependencies(namespace.members),
+                    },
+                )
+
+            # Generate the sub-package __init__.py that pre-imports all
+            # of the sub-modules.
+            subpackage_init_imports.sort()
+            docstring = f"{kind.capitalize()} defined in the '{package_definition.id}' package."
+            render_template(
+                "__init__",
+                os.path.join(subpackage_dir_path, "__init__.py"),
+                {"docstring": docstring, "relImports": subpackage_init_imports},
+            )
+
+    # Package __init__.py
+    render_template(
+        "__init__",
+        os.path.join(package_dir_path, "__init__.py"),
+        {"docstring": package_definition.description, "relImports": package_init_imports},
+    )
+
+
+#
+## Jinja setup
+#
+
+
+def _create_jinja_env(env_globals, logger):
+    """
+    Creates a custom Jinja2 environment with:
+     - A package a loader that automatically finds templates within a
+       'templates' directory in the openassetio_codegen python package.
+     - Updated globals.
+     - Custom filters.
+    """
+    env = jinja2.Environment(loader=jinja2.PackageLoader("openassetio_codegen"))
+    env.globals.update(env_globals)
+    _install_custom_filters(env, logger)
+    return env
+
+
+# Custom filters
+
+
+def _install_custom_filters(environment, logger):
+    """
+    Installs custom filters in to the Jinja template environment that allow
+    data from the model to be conformed to python-specific standards.
+
+    The toPy* methods will log a warning if the string is changed from
+    the input during this process. An error will be raised if this
+    resulted in an empty string.
+    """
+
+    def validate_identifier(string: str, original: str):
+        """
+        Validates some string is a legal python variable name.
+        """
+        if not string.isidentifier():
+            raise ValueError(f"{string}' (from '{original}' is not a valid python identifier.")
+        if keyword.iskeyword(string):
+            raise ValueError(f"{string}' (from '{original}' is a reserved python keyword.")
+
+    def to_py_module_name(string: str):
+        """
+        Conforms the supplied string a legal module name.
+        """
+        # Don't warn for - to _ as it is expected that setuptools
+        # creates a safe name from the package name anyway.
```

Comment:
[question] Is "creates a safe name from the package name" true in all usages? We use it to transform namespaces too. I.e. if a namespace contains hyphens, then no warning is logged, is that expected?

Reply:
Not sure? Given the "automagicness" of package name conformance in python, is it tenable to say that one may expect hyphens to underscores to be a "python thing" in general?

Reply:
OK, that's a reasonable argument.